WeRadiate-ThermoSense.ino

```diff
@@ -405,6 +464,214 @@ void loop(void)
+                gCatena.SafePrintf(
+                        "Si1133:  %u IR, %u White, %u UV\n",
+                        data[0],
+                        data[1],
+                        data[2]
+                        );
+                b.putLux(data[1]);
+                flag |= FlagsSensor3::FlagLux;
+                }
+    
+        /*
+        || Measure and transmit the "water temperature" (OneWire)
+        || tranducer value. This is complicated because we want
+        || to support plug/unplug and the sw interface is not
+        || really hot-pluggable.
+        */
+        
+        /* set D11 high so V_OUT2 is going to be high for onewire sensor */        
+        pinMode(D11, OUTPUT);
+        digitalWrite(D11, HIGH);
+        
+        bool fCompostTemp = checkCompostSensorPresent();
+      
+        if (fCompostTemp)
+                {
+                sensor_CompostTemp.requestTemperatures();
+                float compostTempC = sensor_CompostTemp.getTempCByIndex(0);
+                Serial.print("Compost temperature: "); Serial.print(compostTempC); Serial.println(" C");
+                // transmit the measurement
+                b.putT(compostTempC);
+                flag |= FlagsSensor3::FlagWater;
+                }
+        else if (fHasCompostTemp)
+                {
+                gCatena.SafePrintf("No compost temperature\n");
+                }
+             
+        /* set D11 low to turn off after measuring */        
+//        digitalWrite(D11, LOW);
+        pinMode(D11, INPUT);
+      
+        *pFlag = uint8_t(flag);
+        if (savedLed != LedPattern::Joining)
+                {
+                gLed.Set(LedPattern::Sending);
+                }
+        else
+                {
+                gLed.Set(LedPattern::Joining);
+                }
+        
+        gLoRaWAN.SendBuffer(b.getbase(), b.getn(), sendBufferDoneCb, NULL);
+}
+
+void
+sendBufferDoneCb(
+        void *pContext,
+        bool fStatus
+        )
+        {
+        osjobcb_t pFn;
+
+        gLed.Set(LedPattern::Settling);
+        if (! fStatus)
+                {
+                gCatena.SafePrintf("send buffer failed\n");
+                pFn = txFailedDoneCb;
+                }
+        else
+                {
+                pFn = settleDoneCb;
+                }
+        os_setTimedCallback(
+                &sensorJob,
+                os_getTime()+sec2osticks(CATCFG_T_SETTLE),
+                pFn
+                );
+        }
+
+void
+txFailedDoneCb(
+        osjob_t *pSendJob
+        )
+        {
+        gCatena.SafePrintf("not provisioned, idling\n");
+        gLoRaWAN.Shutdown();
+        gLed.Set(LedPattern::NotProvisioned);
+        }
+
+void settleDoneCb(
+        osjob_t *pSendJob
+        )
+        {
+        // if connected to USB, don't sleep
+        // ditto if we're monitoring pulses.
+	// XXX: do we need this?
+#ifdef USBCON
+        if (fUsbPower || fHasPower1)
+#else
+        if (fHasPower1)
```

Comment:
never are monitoring pulses, so please remove `fHasPower1`.  For the general structure, look at how I do this in the newer sketches. Don't use `#ifdef` around `if`, instead set up a boolean flag that can be tested cleanly.

```c++
#ifdef USBCON
   const bool fCanSleep = ! fUsbPower;
#else
  const bool fCanSlee = true;
#endif

  if (! fCanSleep)
       {
```